pkg/controllermanager/controller/certificatesigningrequest/reconciler_test.go

```diff
@@ -0,0 +1,135 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var _ = Describe("csrReconciler", func() {
+	var (
+		ctx        context.Context
+		fakeClient client.Client
+		restConfig *rest.Config
+
+		csr                *certificatesv1.CertificateSigningRequest
+		reconciler         reconcile.Reconciler
+		privateKey         *rsa.PrivateKey
+		certificateSubject *pkix.Name
+	)
+
+	BeforeEach(func() {
+		restConfig = &rest.Config{}
+		certificatesClient, _ := kubernetesclientset.NewForConfig(restConfig)
+
+		fakeClient = fakeclient.NewClientBuilder().WithScheme(kubernetes.GardenScheme).Build()
+
+		privateKey, _ = secretutils.FakeGenerateKey(rand.Reader, 4096)
+		csr = &certificatesv1.CertificateSigningRequest{
+			TypeMeta: metav1.TypeMeta{Kind: "CertificateSigningRequest"},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "test-csr",
+			},
+			Spec: certificatesv1.CertificateSigningRequestSpec{
+				Usages: []certificatesv1.KeyUsage{
+					certificatesv1.UsageDigitalSignature,
+					certificatesv1.UsageKeyEncipherment,
+					certificatesv1.UsageClientAuth,
+				},
+				SignerName: certificatesv1.KubeAPIServerClientSignerName,
+			},
+		}
+
+		reconciler = &Reconciler{Client: fakeClient, CertificatesClient: certificatesClient, CertificatesAPIVersion: "v1"}
+	})
+
+	It("should return nil because object not found", func() {
+		Expect(fakeClient.Get(ctx, client.ObjectKeyFromObject(csr), &certificatesv1.CertificateSigningRequest{})).To(BeNotFoundError())
+
+		result, err := reconciler.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{Name: csr.Name}})
+		Expect(result).To(Equal(reconcile.Result{}))
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	Context("Final state", func() {
+		BeforeEach(func() {
+			Expect(fakeClient.Create(ctx, csr)).To(Succeed())
+			Expect(fakeClient.Get(ctx, client.ObjectKeyFromObject(csr), csr)).To(Succeed())
+		})
+
+		It("should ignore the csr because certificate is already present in the status field", func() {
+			patch := client.MergeFrom(csr.DeepCopy())
+			csr.Status.Certificate = []byte("test-certificate")
+			Expect(fakeClient.Patch(ctx, csr, patch)).To(Succeed())
+
+			result, err := reconciler.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{Name: csr.Name}})
+			Expect(result).To(Equal(reconcile.Result{}))
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should ignore the csr because csr is already Approved", func() {
+			patch := client.MergeFrom(csr.DeepCopy())
+			csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1.CertificateSigningRequestCondition{
+				Type: certificatesv1.CertificateApproved,
+			})
+			Expect(fakeClient.Patch(ctx, csr, patch)).To(Succeed())
+
+			result, err := reconciler.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{Name: csr.Name}})
+			Expect(result).To(Equal(reconcile.Result{}))
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
+
+	Context("non seedclient csr", func() {
```

Comment:
This test suite misses cases for non-authorized clients and the successful case (i.e., where the CSR is approved)